src/amp-story-embed-manager.js

```diff
@@ -0,0 +1,113 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AmpStoryEmbed} from './amp-story-embed';
+
+/** @const {string} */
+const SCROLL_THROTTLE_MS = 500;
+
+export class AmpStoryEmbedManager {
+  /**
+   * @param {!Window} win
+   * @constructor
+   */
+  constructor(win) {
+    /** @private {!Window} */
+    this.win_ = win;
+  }
+
+  /**
+   * Calls layoutCallback on the embed when it is close to the viewport.
+   * @param {!AmpStoryEmbed} embedImpl
+   * @private
+   */
+  layoutEmbed_(embedImpl) {
+    if (!IntersectionObserver || this.win_ !== this.win_.parent) {
+      this.layoutFallback_(embedImpl);
+      return;
+    }
+
+    const intersectingCallback = entries => {
+      entries.forEach(entry => {
+        if (!entry.isIntersecting) {
+          return;
+        }
+        embedImpl.layoutCallback();
+      });
+    };
+
+    const observer = new IntersectionObserver(intersectingCallback, {
+      rootMargin: '100%',
```

Comment:
`rootMargin` will not work inside iframes. /cc @choumx

Reply:
Yep, I wrote a fallback when IO is not supported or if it's inside an iframe. See line 38.